source/lib/JMSLab/builders/build_latex.py

```diff
@@ -44,3 +46,40 @@ def add_call_args(self):
                                   os.path.normpath(self.log_file))
         self.call_args = args
         return None
+     
+    def add_bib_name(self, target):
+        if bool(target):
+            target = misc.make_list_if_string(target)
+            target_file = str(target[0]).split(".")[0]
+        else:
+            target_file = ''
+        self.bib_file = target_file
+        return None
+    
+    def do_call(self):
+        '''
+        Acutally execute the system call attribute.
+        Raise an informative exception on error.
+        '''
+        self.bibtex_executable  = 'bibtex'
+        self.bibtex_system_call = '%s %s' % (self.bibtex_executable, self.bib_file)
+        
+        traceback = ''
+        raise_system_call_exception = False
+        try:
+            subprocess.check_output(self.system_call, shell = True, stderr = subprocess.STDOUT)
+            subprocess.check_output(self.bibtex_system_call, shell = True, stderr = subprocess.STDOUT)
```

Comment:
I think we should have a conditional logic that would make the bibtex call only when a bibliography is used.

Reply:
@veli-m-andirin This would not be sufficient as `bibtex` also fails when no citations are defined in the main text (even if there is a bibliography file referenced).

Reply:
I think a structure where

1) bibtex is not called when a bibliography is not referenced
2) bibtex is called when a bibliography is referenced

would be fine. What you say would fall into the second case. Even though it would raise an error if citations are absent, I think the solution there would be to modify the source file. (If someone references a bibliography, then it is up to them to use it, and if they don't use it, they can just remove it.) I can also imagine a case where we also confirm that citations are indeed made, but this might be more than needed.

Reply:
@veli-m-andirin @mcaceresb In 0ddc2c6 I modified the builder so that it only calls bibtex if a .bib file is included in the sources in scons. If there is a .bib file in sources, it runs the normal pdflatex -> bibtex -> pdflatex -> pdflatex. If there is no .bib file, it runs pdflatex twice as per [this comment](https://github.com/JMSLab/Template/issues/42#issuecomment-961205623). 

@jmshapir fyi.

Reply:
@rcalvo12 thanks! Made some minor comments.

As for running it twice, per the comment, 3 is safer, though admittedly it's a bit of an edge case. I've added a simple example in 077b56c5e5cbae266d803a24834be67e568a06d7 where running `pdflatex` twice does not give the right reference.

@veli-m-andirin fyi in case you have thoughts on this point.

Reply:
@mcaceresb, if you don't think there are any serious drawbacks to running three times (I don't think runtime qualifies as a serious drawback), then I definitely agree with doing what is safer.

Reply:
@mcaceresb @veli-m-andirin I implemented a third run in 7bfca71, forgot to leave a note here about it.

Reply:
@veli-m-andirin I believe this conversation can be resolved, let me know if there is anything else here that needs to happen.